// @cucumber/query/dist/src/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assert = void 0;
exports.comparatorById = comparatorById;
exports.comparatorBy = comparatorBy;
exports.comparatorByStatus = comparatorByStatus;
const messages_1 = require("@cucumber/messages");
function comparatorById(a, b) {
    return comparatorBy(a, b, "id");
}
function comparatorBy(a, b, key) {
    if (a[key] < b[key]) {
        return -1;
    }
    if (a[key] > b[key]) {
        return 1;
    }
    return 0;
}
function comparatorByStatus(a, b) {
    if (ordinal(a.status) < ordinal(b.status)) {
        return -1;
    }
    if (ordinal(a.status) > ordinal(b.status)) {
        return 1;
    }
    return 0;
}
function ordinal(status) {
    return [
        messages_1.TestStepResultStatus.UNKNOWN,
        messages_1.TestStepResultStatus.PASSED,
        messages_1.TestStepResultStatus.SKIPPED,
        messages_1.TestStepResultStatus.PENDING,
        messages_1.TestStepResultStatus.UNDEFINED,
        messages_1.TestStepResultStatus.AMBIGUOUS,
        messages_1.TestStepResultStatus.FAILED,
    ].indexOf(status);
}
exports.assert = {
    ok(target, message) {
        if (!target) {
            throw new Error(message);
        }
    }
};
